code/_onclick/ai.dm

```diff
@@ -148,20 +149,23 @@
 		shock_perm(usr)
 	else
 		shock_restore(usr)
+	ui_update()
```

Comment:
Ui update should be handled in shock perm and shock restore I think

Reply:
I could change that, but that'd introduce redundant updates, because shock_perm is used in ui_act as well. With how ui_update works right now, it updates the next process, whereas ui_act updates immediately, so if you ui_update inside ui_act, then trigger an update from ui_act, it'll update twice.

I'm thinking now, such issues could be avoided if doing any update set `ui.needs_update` to false, which would prevent such redundant updates, with no drawback I can see.

Reply:
I moved the ui_update calls deeper in where suggested, gonna resolve all comments regarding that now. I think `ui.needs_update` should be getting reset on any UI update, not just ones *caused by* `needs_update`, since that would help reduce the amount of updates done while cramming `ui_update` calls into things called by `ui_act`.